src/components/ArticleView.jsx

```diff
@@ -1,24 +1,46 @@
-import { fetchArticleByID } from '../data-fetching';
+import { fetchArticleByID, patchVotes } from '../utils/api';
 import { useParams } from 'react-router-dom';
 import { useState, useEffect } from 'react';
 import { Typography } from '@mui/material';
+import { IconButton } from '@material-ui/core';
+import { ThumbUp } from '@mui/icons-material';
 
 export default function ArticleView() {
   const { article_id } = useParams();
   const [article, setArticle] = useState({});
+  const [votes, setVotes] = useState(0);
+  const [hasVoted, setHasVoted] = useState(false);
 
   useEffect(() => {
     fetchArticleByID(article_id).then((article) => {
       setArticle(article[0]);
     });
   }, [article_id]);
 
+  const incrementVotes = () => {
+    let incOrDec = 1;
+    if (hasVoted) incOrDec = -1;
+
+    setVotes((currVotes) => {
+      return currVotes + incOrDec;
+    });
+    patchVotes(article.article_id, incOrDec).catch(() => {
+      setVotes((currVotes) => {
+        return currVotes + incOrDec;
+      });
+    });
+    setHasVoted((curr) => !curr);
+  };
+
```

Comment:
The optimistic rendering looks great, just a point on implementation (and another point on the error handling!)

1. Your boolean in state to check whether someone `hasVoted` is probably not necessary. The `votes` state that you're using for the optimistic rendering is only ever `0` or `1` which makes sense! But this means you can effectively use whether that value is `0` or `1` to tell whether someone has voted and you don't need that extra value in state.
2. Your error handling is almost perfect, you just need to change line 29 to be `return currVotes - incOrDec` and then it'll be great 👍

Reply:
SUAVE 🕺🏾 Thanks Liam!